File: MlpNetwork/activationfunction.cpp
#include "stdafx.h"
#include "activationfunction.h"

namespace mlp_network
{
	ActivationFunction::ActivationFunction(const ActivationFunction::Type &type)
	{
		switch (type)
		{
		case Type::UNIPOLAR_SIGMOID:
			function_ = [](double x) { return 1 / (1 + exp(-x)); };
			derivative_ = [](double fX) { return fX * (1 - fX); };
			break;
		case Type::BIPOLAR_SIGMOID:
			function_ = [](double x) { return tanh(x); };
			derivative_ = [](double fX) { return 1 - fX * fX; };
			break;
		case Type::SINUSOID:
			function_ = [](double x) { return sin(x); };
			derivative_ = [](double fX) { return sqrt(1 - fX * fX); };
			break;
		case Type::LINEAR:
			function_ = [](double x) { return x; };
			derivative_ = [](double fX) { return 1.0; };
			break;
		default:
			function_ = [](double x) { return 0.0; };
			derivative_ = [](double fX) { return 0.0; };
			break;
		}
	}

	double ActivationFunction::value(double x) const
	{
		return function_(x);
	}

	double ActivationFunction::derivative(double fX) const
	{
		return derivative_(fX);
	}
}

File: MlpPredictorCuda/MlpPredictor/ActivationFunction.cs
using System;

namespace MlpPredictor
{
    public enum ActivationFunctionType
    {
        UnipolarSigmoid = 0,
        BipolarSigmoid = 1,
        Sinusoid = 2,
        Linear = 3
    }

    [Serializable]
    public class ActivationFunction
    {
        private Func<float, float> function;
        private Func<float, float> derivative;

        public ActivationFunction(ActivationFunctionType type)
        {
            switch (type)
            {
                case ActivationFunctionType.UnipolarSigmoid:
                    function = x => 1.0f / (1.0f + (float)Math.Exp(-x));
                    derivative = fX => fX * (1.0f - fX);
                    break;
                case ActivationFunctionType.BipolarSigmoid:
                    function = x => (float)Math.Tanh(x);
                    derivative = fX => 1.0f - fX * fX;
                    break;
                case ActivationFunctionType.Sinusoid:
                    function = x => (float)Math.Sin(x);
                    derivative = fX => (float)Math.Sqrt(1.0f - fX * fX);
                    break;
                case ActivationFunctionType.Linear:
                    function = x => x;
                    derivative = fX => 1.0f;
                    break;
                default:
                    throw new ArgumentException("Неизвестная функция активации.", "type");
            }

            Type = type;
        }

        public ActivationFunctionType Type { get; private set; }

        public float Value(float argument)
        {
            return function(argument);
        }

        public float Derivative(float functionValue)
        {
            return derivative(functionValue);
        }
    }
}
